lib/Stats.js

```diff
@@ -1,24 +1,55 @@
 /*
-	MIT License http://www.opensource.org/licenses/mit-license.php
-	Author Tobias Koppers @sokra
+MIT License http://www.opensource.org/licenses/mit-license.php
+Author Tobias Koppers @sokra
 */
 "use strict";
 
 const RequestShortener = require("./RequestShortener");
 const SizeFormatHelpers = require("./SizeFormatHelpers");
 const formatLocation = require("./formatLocation");
 const identifierUtils = require("./util/identifier");
+const Tapable = require("tapable").Tapable;
+const SyncHook = require("tapable").SyncHook;
+const HookMap = require("tapable").HookMap;
+const SyncWaterfallHook = require("tapable").SyncWaterfallHook;
 
 const optionsOrFallback = (...args) => {
 	let optionValues = [];
 	optionValues.push(...args);
 	return optionValues.find(optionValue => typeof optionValue !== "undefined");
 };
 
-class Stats {
+const isTypeOf = (variable, type) => {
+	if (typeof variable === type) return true;
+	return false;
+};
+
+const assertTypeOf = (variable, type) => {
+	if (typeof variable === type || typeof variable === "undefined")
+		return variable;
+	throw new Error(
+		"Expected " +
+			variable +
+			" to be of type " +
+			type +
+			"but provided with type" +
+			typeof variable
+	);
+};
+
+class Stats extends Tapable {
 	constructor(compilation) {
+		super();
 		this.compilation = compilation;
 		this.hash = compilation.hash;
+		this.hooks = {
+			willCheckErrors: new SyncHook(),
+			willCheckWarnings: new SyncHook(),
+			warning: new SyncHook(),
+			error: new SyncHook(),
+			setDefault: new HookMap(() => new SyncWaterfallHook(["lastPluginValue"]))
```

Comment:
I'm not convinced by this hook. Using a `HookMap` is not user-friendly if a plugin wants to register multiple values. What do you think about using a `SyncHook` that returns an object which would be merged with the provided options? The current options could be passed to the hook callback so the defaults could be adjusted.

Hook would be defined this way:

```js
getDefaults: new SyncHook(["options"]),
```

And the stats options could be merged:

```js
const statsOptions = Object.assign(
  {},
  options,
  this.hooks.getDefaults.call(
    Object.assign({}, options) // prevent mutation
  )
)

Reply:
@ooflorent, I used `HookMap` because we could check for individual values, their types and whether they be allowed or not. I agree assigning multiple values is not user friendly, however I think plugin authors can use a loop to tap into setdefaults for most of their values. ( Just like [NoStatsPreset.js](#diff-c9ae2b87a5e7987bda17ba294a8a314e) ), That way, they can use an object, and use for-in/ Object.for and then tap using loop. This will also keep the purpose of PR to validate the options supplied. Thoughts?